controllers/notificationpolicy_controller.go

```diff
@@ -135,12 +165,104 @@ func (r *GrafanaNotificationPolicyReconciler) Reconcile(ctx context.Context, req
 		return ctrl.Result{}, fmt.Errorf("failed to apply to all instances: %v", applyErrors)
 	}
 
+	if len(mergedRoutes) > 0 {
+		status := statusDiscoveredRoutes(mergedRoutes)
+		notificationPolicy.Status.DiscoveredRoutes = &status
+	}
+
+	if err := r.updateNotificationPolicyRoutesStatus(ctx, notificationPolicy, mergedRoutes); err != nil {
+		r.Log.Error(err, "failed to add merged events to routes")
+	}
+
+	if notificationPolicy.Spec.Route.IsRouteSelectorMutuallyExclusive() {
+		meta.RemoveStatusCondition(&notificationPolicy.Status.Conditions, conditionRoutesIgnoredDueToRouteSelector)
+	} else {
+		setRoutesIgnoredDueToRouteSelectorCondition(&notificationPolicy.Status.Conditions, notificationPolicy.Generation)
+	}
+
 	condition := buildSynchronizedCondition("Notification Policy", conditionNotificationPolicySynchronized, notificationPolicy.Generation, applyErrors, len(instances))
 	meta.SetStatusCondition(&notificationPolicy.Status.Conditions, condition)
 
 	return ctrl.Result{RequeueAfter: notificationPolicy.Spec.ResyncPeriod.Duration}, nil
 }
 
+// assembleNotificationPolicyRoutes iterates over all routeSelectors transitively.
+// returns an assembled GrafanaNotificationPolicy as well as a list of all merged routes.
+// it ensures that there are no reference loops when discovering routes via labelSelectors
+
+func assembleNotificationPolicyRoutes(ctx context.Context, k8sClient client.Client, namespace *string, notificationPolicy *grafanav1beta1.GrafanaNotificationPolicy) (*grafanav1beta1.GrafanaNotificationPolicy, []*v1beta1.GrafanaNotificationPolicyRoute, error) {
+	assembledPolicy := notificationPolicy.DeepCopy()
```

Comment:
Same here, the `DeepCopy` is currently unnecessary.

Reply:
fixed in https://github.com/grafana/grafana-operator/pull/1800/commits/36c1bede6b24d1c353a0c0917eecd6f497095a1e